File: ticket_office/java/src/main/java/fr/arrolla/trainreservation/domain/Seat.java
package fr.arrolla.trainreservation.domain;

public class Seat {
  private final SeatNumber number;
  private final CoachID coach;
  private String bookingReference;

  public Seat(SeatNumber number, CoachID coach, String bookingReference) {
    this.number = number;
    this.coach = coach;
    this.bookingReference = bookingReference;
  }

  public Seat(SeatNumber number, CoachID coach) {
    this(number, coach, null);
  }

  public SeatID id() {
    return new SeatID(number, coach);
  }

  public void book(String bookingReference) {
    if (this.isFree()) {
      this.bookingReference = bookingReference;
    }

    if (this.bookingReference.equals(bookingReference)) {
      return;
    }
    
    String message = String.format("Seat '%s' already booked with reference '%s', but trying to book with '%s",
      this.number,
      this.bookingReference,
      bookingReference
    );

    throw new AlreadyBookedException(message);
  }

  public boolean isFree() {
    return this.bookingReference == null;
  }

  public SeatNumber number() {
    return number;
  }

  public CoachID coach() {
    return coach;
  }

  public String bookingReference() {
    return bookingReference;
  }
}

File: ts/ticket-office/domain/Seat.ts
export class SeatNumber {
  value: string

  constructor(value: string) {
    this.value = value
  }
}

export class CoachId {
  value: string

  constructor(value: string) {
    this.value = value
  }
}

export class SeatId {
  number: SeatNumber
  coach: CoachId

  constructor(number: SeatNumber, coach: CoachId) {
    this.number = number
    this.coach = coach
  }

  toString(): string {
    return this.number.value + this.coach.value
  }
}


export default class Seat {
  id: SeatId
  booking_reference: string

  constructor(id: SeatId, bookingReference: string) {
    this.id = id
    this.booking_reference = bookingReference
  }

  get coach(): string {
    return this.id.coach.value
  }

  get number(): string {
    return this.id.number.value
  }
}
